espnet/st/pytorch_backend/st.py

```diff
@@ -0,0 +1,632 @@
+#!/usr/bin/env python
+
+# Copyright 2019 Kyoto University (Hirofumi Inaguma)
+#  Apache 2.0  (http://www.apache.org/licenses/LICENSE-2.0)
+
+
+import copy
+import json
+import logging
+import math
+import os
+import sys
+
+from chainer.datasets import TransformDataset
+from chainer import training
+from chainer.training import extensions
+from chainer.training.updater import StandardUpdater
+import numpy as np
+from tensorboardX import SummaryWriter
+import torch
+from torch.nn.parallel import data_parallel
+
+from espnet.asr.asr_utils import adadelta_eps_decay
+from espnet.asr.asr_utils import add_results_to_json
+from espnet.asr.asr_utils import CompareValueTrigger
+from espnet.asr.asr_utils import get_model_conf
+from espnet.asr.asr_utils import restore_snapshot
+from espnet.asr.asr_utils import snapshot_object
+from espnet.asr.asr_utils import torch_load
+from espnet.asr.asr_utils import torch_resume
+from espnet.asr.asr_utils import torch_snapshot
+import espnet.lm.pytorch_backend.extlm as extlm_pytorch
+from espnet.nets.asr_interface import ASRInterface
+from espnet.nets.mt_interface import MTInterface
+from espnet.nets.pytorch_backend.e2e_asr import pad_list
+import espnet.nets.pytorch_backend.lm.default as lm_pytorch
+from espnet.nets.pytorch_backend.streaming.segment import SegmentStreamingE2E
+from espnet.nets.pytorch_backend.streaming.window import WindowStreamingE2E
+from espnet.nets.st_interface import STInterface
+from espnet.utils.deterministic_utils import set_deterministic_pytorch
+from espnet.utils.dynamic_import import dynamic_import
+from espnet.utils.io_utils import LoadInputsAndTargets
+from espnet.utils.training.batchfy import make_batchset
+from espnet.utils.training.iterators import ShufflingEnabler
+from espnet.utils.training.iterators import ToggleableShufflingMultiprocessIterator
+from espnet.utils.training.iterators import ToggleableShufflingSerialIterator
+from espnet.utils.training.tensorboard_logger import TensorboardLogger
+from espnet.utils.training.train_utils import check_early_stop
+from espnet.utils.training.train_utils import set_early_stop
+
+from espnet.asr.pytorch_backend.asr import CustomEvaluator
+from espnet.asr.pytorch_backend.asr import load_trained_model
+
+import matplotlib
+matplotlib.use('Agg')
+
+if sys.version_info[0] == 2:
+    from itertools import izip_longest as zip_longest
+else:
+    from itertools import zip_longest as zip_longest
+
+
+class CustomUpdater(StandardUpdater):
```

Comment:
Is it possible to reuse this Updater and Converter from ASR?

Reply:
I'm using multitask learning with ASR. Therefore I made this PR.

Reply:
OK. Now I can see the diference in CustomConverter. However, I expect that we can extend CustomConverter to support multiple output json. This strategy is hard to maintain for us because the code has many duplicated parts.

I also expect that CustomUpdater can be unified.